### ISSUE TITLE
Module:sca Line seize does not release after some 'BYE's

### ISSUE THREAD

**Issue by TheGrandWazoo** [kamailio/kamailio#341]
I have a situation that I can't get my head around. Every so often I have a SCA seized and I need to manually release it. Here is the log of the sca and bye from the event logs.

messages-20150920:Sep 18 14:35:28 sip-registrar /usr/sbin/kamailio[7608]: ERROR: sca [sca_call_info.c:1580]: sca_call_info_bye_handler(): sca_call_info_bye_handler: sip:Mindfield@kamailio.newgarllc.com dialog leg 96b9ad37-9a7fa462-15c35d3@192.168.1.117;1c389191866 is not active

messages-20150920:Sep 18 14:35:28 sip-registrar /usr/sbin/kamailio[7610]: NOTICE: acc [acc.c:317]: acc_log_request(): ACC: transaction answered: timestamp=1442601328;method=BYE;from_tag=76C3EC1F-114D759A;to_tag=1c389191866;call_id=96b9ad37-9a7fa462-15c35d3@192.168.1.117;code=200;reason=OK;src_user=Mindfield;src_domain=kamailio.newgarllc.com;src_ip=192.168.1.40;dst_ouser=Mindfield;dst_user=Mindfield;dst_domain=192.168.1.117

And the kamcmd showing the seize line

sip:Mindfield@kamailio.newgarllc.com 1 active 1442601278 sip:Mindfield@192.168.1.117 sip:1030@192.168.1.40:5060 96b9ad37-9a7fa462-15c35d3@192.168.1.117 76C3EC1F-114D759A 1c389191866

I am running Kamailio 4.2.6. The phones are Polycom VVX 310/410 running 5.3.1. And the PSTN gateway is an Audiocodes Mediant 1000 running 6.60A.292.001.

The call volume is light so I don't think there is a lot of traffic but there is about 16 phones one SCA. It is not happening all time. Could be a couple days that nothing happens and then there could be a couple days that it happens multiple times a day. I can't find a common element that would be causing this unless the 'bye' is closing the call before the SCA has a chance to do something with it.

Thanks
Kev

**Comment by kamailio-sync** [comment]
If you can get a clean packet capture, feel free to send it to me off-list.

If you can't, then perhaps you can dig all of the Kamailio logs out for 
those call IDs?

-- Jorj

**Comment by TheGrandWazoo** [comment]
It will be hard to get a packet capture but I can give you what I have in the logs. I will also get homer going which should give us a better capture due to it sporadicness.
Where would you like me to send it too?

**Comment by kamailio-sync** [comment]
You can send it directly to me off-list.

-- Jorj

**Comment by TheGrandWazoo** [comment]
Jorj,

What is the email address to send you the capture and SCA output. I think I see the issue but don't know why it happens?

**Comment by TheGrandWazoo** [comment]
Ill try this...hope it works. 

sip:Mindfield@kamailio.newgarllc.com 1 active 1444143590 sip:Mindfield@192.168.1.85 sip:1030@192.168.1.40:5060 e90b49fb-d94a179a-a738a3@192.168.1.85 7EF5202B-6BC5E1EA 1c836141488 

Thanks, 




Kevin Scott Adams 
KSA Technologies, LLC 
4625 Sundew Street 
Commerce Township, MI. 48382 


http://www.ksatechnologies.com 
kevin.adams@ksatechnologies.com 
P: +1 248 521 8203 
----- Original Message -----

**Comment by TheGrandWazoo** [comment]
Jorj,
The issue seems to be only with the Polycom UA making a call outbound. I can not find an issue with and inbound call. The Call-ID of the seize/appearance seems to be missing some characters on the prefix of the at-sign in relations to the real Call-ID of the INVITE/BYE. 

Just my observation.

Kev

**Comment by JorjBauer** [comment]
Apologies for being unresponsive; I've recently left Penn.

I'll need more detail, unfortunately. Can you send any log messages 
relating to that Call ID?

e90b49fb-d94a179a-a738a3@192.168.1.85

-- Jorj

**Comment by TheGrandWazoo** [comment]
Jorj, 

I do not have that log file any more. I will need to look for a new seize and give you that info. 

Sorry about that, 




Kevin Scott Adams 
KSA Technologies, LLC 
4625 Sundew Street 
Commerce Township, MI. 48382 


http://www.ksatechnologies.com 
kevin.adams@ksatechnologies.com 
P: +1 248 521 8203 
----- Original Message -----

**Comment by TheGrandWazoo** [comment]
Jorj,
It happened again but the logs do not show any reference to the Call-ID except the <BYE>. I can not find the invite for that call in the logs.
I have a Homer capture of the <INVITE> to the <BYE>.

I'm going to see on my test system if I can find out why we might be missing these <INVITE>s.

**Comment by miconda** [comment]
Reopen if this issue is still actual and new details can be provided.

**Comment by TheGrandWazoo** [comment]
Yes, it is still happening under 4.2 at the current git patch. 

Phones are Polycom VVX310/410 at 5.3.2 and I have Homer running to capture. 

Problem is only outbound from Polycom to an outside line via an Audiocode Mediant 1000 with the phone(s) having 2 different SCA's and 2 to 3 line appearances per SCA. 

Thanks, 




Kevin Scott Adams 




4625 Sundew Street 
Commerce Township, MI. 48382 
Phone: +1 248 521 8203 
Email: kevin.adams@ksatechnologies.com 
Web: www.ksatechnologies.com 




----- Original Message -----

**Comment by TheGrandWazoo** [comment]
Is there a committer/developer for this module?

**Comment by TheGrandWazoo** [comment]
Typical log entry under the [SCA] route for this situation...

Jun  6 11:14:28 sip-registrar /usr/sbin/kamailio[15361]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=984 a=16 n=if
Jun  6 11:14:28 sip-registrar /usr/sbin/kamailio[15361]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=974 a=25 n=is_method
Jun  6 11:14:28 sip-registrar /usr/sbin/kamailio[15361]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=990 a=16 n=if
Jun  6 11:14:28 sip-registrar /usr/sbin/kamailio[15361]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=984 a=25 n=is_method
Jun  6 11:14:28 sip-registrar /usr/sbin/kamailio[15361]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=990 a=24 n=sca_call_info_update
Jun  6 11:14:28 sip-registrar /usr/sbin/kamailio[15361]: DEBUG: sca [sca_util.c:53]: sca_get_msg_contact_uri(): Empty Contact header
Jun  6 11:14:28 sip-registrar /usr/sbin/kamailio[15361]: DEBUG: sca [sca_util.c:53]: sca_get_msg_contact_uri(): Empty Contact header
Jun  6 11:14:28 sip-registrar /usr/sbin/kamailio[15361]: ERROR: sca [sca_call_info.c:1580]: sca_call_info_bye_handler(): sca_call_info_bye_handler: sip:ParksAndRec@kamailio.newgarllc.com dialog leg 170216930662016111325@192.168.1.40;584cf90b-57a7-4543-bfda-fd7cd71ca84c is not active
Jun  6 11:14:28 sip-registrar /usr/sbin/kamailio[15361]: ERROR: sca [sca_call_info.c:2032]: sca_call_info_update(): Failed to update Call-Info state for
Jun  6 11:14:28 sip-registrar /usr/sbin/kamailio[15361]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=993 a=2 n=return

The [SCA] route is the one in the modules readme or website doc.

**Comment by TheGrandWazoo** [comment]
Another seized via kamcmd sca.all_appearances

Saturday, June 04, 2016 1:04:30 PM
sip:RepublicTavern@kamailio.newgarllc.com 1 active 1465059870 sip:RepublicTavern@192.168.1.111 sip:1030@192.168.1.40:5060 fe298e35-e351232c-867b4191@192.168.1.111 F8EA101D-EEDFCE4 1c1260378071

Log on the siezed....
Jun  4 13:05:40 sip-registrar /usr/sbin/kamailio[5508]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=983 a=16 n=if
Jun  4 13:05:40 sip-registrar /usr/sbin/kamailio[5508]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=973 a=25 n=is_method
Jun  4 13:05:40 sip-registrar /usr/sbin/kamailio[5508]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=989 a=16 n=if
Jun  4 13:05:40 sip-registrar /usr/sbin/kamailio[5508]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=983 a=25 n=is_method
Jun  4 13:05:40 sip-registrar /usr/sbin/kamailio[5508]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=989 a=24 n=sca_call_info_update
Jun  4 13:05:40 sip-registrar /usr/sbin/kamailio[5508]: DEBUG: sca [sca_util.c:53]: sca_get_msg_contact_uri(): Empty Contact header
Jun  4 13:05:40 sip-registrar /usr/sbin/kamailio[5508]: DEBUG: sca [sca_util.c:53]: sca_get_msg_contact_uri(): Empty Contact header
Jun  4 13:05:40 sip-registrar /usr/sbin/kamailio[5508]: ERROR: sca [sca_call_info.c:1580]: sca_call_info_bye_handler(): sca_call_info_bye_handler: sip:RepublicTavern@kamailio.newgarllc.com dialog leg fe298e35-e351232c-867b4191@192.168.1.111;1c1260378071 is not active
Jun  4 13:05:40 sip-registrar /usr/sbin/kamailio[5508]: ERROR: sca [sca_call_info.c:2032]: sca_call_info_update(): Failed to update Call-Info state for
Jun  4 13:05:40 sip-registrar /usr/sbin/kamailio[5508]: exec: *** cfgtrace:request_route=[SCA] c=[//etc/kamailio/kamailio.cfg] l=992 a=2 n=return

**Comment by miconda** [comment]
@JorjBauer used to maintain the module at some point, but I think he changed positions and not sure if he is still involved in some work with the module.

**Comment by TheGrandWazoo** [comment]
Good day.
I think I have a lead on this to help come up with a resolution. I might be able to do it myself but I need to really dig into the code and kamailio to understand it better.
First, I was using the sip_trace module along with Homer to gather the info, but it did seem incomplete. The work of the SCA module is not picked up by the sip_trace module. I believe because it if doing instead script work and sending its own packets, which would not be picked up by the script anymore. I finally used sngrep (referred by Javan Strika who is also having the issue) and was able to see the entire conversations for the SCA's. But the sngrep in a "HEP" and "non verbose" mode seems to have a memory leak and will crash kamailio due to no more memory. So, I am now turning to captagent to get a more stable system to get better traces for this issue.

Now for the SCA problem. It seems that the unreleased "seized event" resides with the "held" event and the reinvite/invite that occurs to kamailio. It seems to get lost, hungup. pushed around or jacked up in the 'route[WITHINDLG]' section of the script and the reinvites/invites will keep happening until a timeout occurs or the sometimes I have seen the 404 Not here error occur.

I know this is not much to go on until I upload a good trace but maybe a brainiac might go...Oh, I know that problem and quicker fix.

Using Kamailio 4.2.7 with latest patches. CentOS 6.7-latest. One test bed is a mix of 2 Polycom VVX [34]10 with 5.4.4 and 1 Polycom 650 with 3.3.5; Another is a mix of 14 Polycoms VVX [34]10 with 5.3.2, another is a mix of 4 Polycom VVX [34]10 with 5.3.2

**Comment by TheGrandWazoo** [comment]
Here is a pcap file of the SCA failing on a HOLD. Seems to be having an issue with the [WITHINDLG] route. This is also try the a regular hangup (BYE) gets the some results with the 404 Not Here. Went all the way back to 4.1 with the same results.
Maybe you can shed some light on the subject of WHY this behavior happens.
[sca-hold-failed-withindlg.zip](https://github.com/kamailio/kamailio/files/374962/sca-hold-failed-withindlg.zip)

**Comment by TheGrandWazoo** [comment]
Hello again,
I have done the following to the [WITHINDLG] route and seems to have made the system to seem to function normally with SCA. The BYE actually now works. Hold works, but I can't pickup from another phone (working on that) and I can transfer a call to a private extension.

So between...

    if (!has_totag()) {
        return;
    }

and this

    # sequential request within a dialog should
    # take the path determined by record-routing
    if (loose_route()) {

I added the following

    if (is_method("PRACK|ACK") && src_ip==<pstn gateway>) {
        route(RELAY);
        exit;
    }

    if (is_method("INVITE|BYE|REFER") && is_present_hf("Call-Info")) {
        route(RELAY);
        exit;
    }

    if (is_method("ACK")) {
        route(RELAY);
        exit;
    }

So I guess I need to understand this loose_route a bit more and why it is causing the issues with the SCA module functions and methods.

I am also wondering if this [WITHINDLG] was part of the 3.x config file. When I upgraded to 4.x I used that one until I saw changes to the 4.x config file. Just shooting from the hip.

Thanks

**Comment by TheGrandWazoo** [comment]
Good day to people following this.
First please disregard  the previous posts with the in dialog issues. I had the IP address of the Audiocodes Gateway in my domain table by accident. Removing that cleared all the in dialog issues up. I will be deleting those posts.

But, this issue still persists and I can confirm and duplicate the issue now that has been haunting me (and others) for quite some time. This issue is when a call is placed on HOLD and another phone picks up the 'held' call...all seems fine until the hangup. This is when the line is still considered 'active' by the SCA module. The current UA (Polycom) seems to have released it but all other phones have the 'appearance' still active. Within 15 to 20 seconds the UA that was part of the released call receives the 'active' appearance. It can be duplicated every time.

**Comment by TheGrandWazoo** [comment]
If people are following this issue thread, I have decided to forked the kamailio repo to my Github account and I am actively working on fixing the issue(s) i have encountered, cleaning up code and hoping to enhance the code to make the module better. The fork is at [https://github.com/TheGrandWazoo/kamailio](https://github.com/TheGrandWazoo/kamailio) and I will be committing some changes soon.
Disclaimer...I am not an internal kamailio or SCA expert so I am learning as I go and there are a lot of DEBUG statements and XXX notes I have added to help me and others along if they so do wish. I have changed code that has broken other functionality of the module so it is not perfect yet. There has been some changes to the Polycom firmware that I have seen in traces that need to be considered for future enhancements that I hope to incorporate into the module.

Thanks
